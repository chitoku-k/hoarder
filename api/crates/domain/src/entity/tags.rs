use std::collections::BTreeSet;

use chrono::{DateTime, Utc};
use derive_more::{Constructor, Deref, Display, From};
use serde::Serialize;
use uuid::Uuid;

#[derive(Clone, Copy, Debug, Default, Deref, Display, Eq, From, Hash, Ord, PartialEq, PartialOrd)]
pub struct TagId(Uuid);

#[derive(Clone, Constructor, Copy, Debug, Eq, From, PartialEq)]
pub struct TagDepth {
    parent: u32,
    children: u32,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct Tag {
    pub id: TagId,
    pub name: String,
    pub kana: String,
    pub aliases: AliasSet,
    pub parent: Option<Box<Self>>,
    pub children: Vec<Self>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl TagId {
    pub const fn root() -> Self {
        Self(Uuid::nil())
    }

    pub const fn is_root(&self) -> bool {
        self.0.is_nil()
    }
}

impl TagDepth {
    pub const fn parent(&self) -> u32 {
        self.parent
    }

    pub const fn children(&self) -> u32 {
        self.children
    }

    pub const fn has_parent(&self) -> bool {
        self.parent > 0
    }

    pub const fn has_children(&self) -> bool {
        self.children > 0
    }
}

#[derive(Clone, Constructor, Debug, Default, Deref, Eq, PartialEq, Serialize)]
pub struct AliasSet(BTreeSet<String>);

impl AliasSet {
    pub fn add_all<T>(&mut self, aliases: T)
    where
        T: IntoIterator<Item = String>,
    {
        for alias in aliases {
            self.0.insert(alias);
        }
    }

    pub fn remove_all<T>(&mut self, aliases: T)
    where
        T: IntoIterator<Item = String>,
    {
        for alias in aliases {
            self.0.remove(&alias);
        }
    }

    pub fn into_inner(self) -> BTreeSet<String> {
        self.0
    }
}

impl From<Vec<String>> for AliasSet {
    fn from(v: Vec<String>) -> Self {
        Self::new(v.into_iter().collect())
    }
}

impl From<AliasSet> for Vec<String> {
    fn from(v: AliasSet) -> Self {
        v.0.into_iter().collect()
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn tag_id_root() {
        let actual = TagId::root();

        assert!(actual.0.is_nil());
    }

    #[test]
    fn tag_id_is_root() {
        let actual = TagId::root();

        assert!(actual.is_root());
    }

    #[test]
    fn tag_depth_parent() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.parent();

        assert_eq!(actual, 1);
    }

    #[test]
    fn tag_depth_children() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.children();

        assert_eq!(actual, 2);
    }

    #[test]
    fn tag_depth_has_parent() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.has_parent();

        assert!(actual);
    }

    #[test]
    fn tag_depth_has_no_parent() {
        let depth = TagDepth::new(0, 2);
        let actual = depth.has_parent();

        assert!(!actual);
    }

    #[test]
    fn tag_depth_has_children() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.has_children();

        assert!(actual);
    }

    #[test]
    fn tag_depth_has_no_children() {
        let depth = TagDepth::new(1, 0);
        let actual = depth.has_children();

        assert!(!actual);
    }

    #[test]
    fn alias_set_add_all() {
        let mut actual = AliasSet::new(BTreeSet::from(["baz".to_string()]));
        actual.add_all(["foo".to_string(), "bar".to_string()]);

        assert_eq!(actual, AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()])));
    }

    #[test]
    fn alias_set_remove_all() {
        let mut actual = AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()]));
        actual.remove_all(["foo".to_string(), "bar".to_string()]);

        assert_eq!(actual, AliasSet::new(BTreeSet::from(["baz".to_string()])));
    }

    #[test]
    fn alias_set_into_inner() {
        let set = AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()]));
        let actual = set.into_inner();

        assert_eq!(actual, BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()]));
    }

    #[test]
    fn alias_set_from_strings() {
        let actual = AliasSet::from(vec!["foo".to_string(), "bar".to_string(), "baz".to_string()]);

        assert_eq!(actual, AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()])));
    }
}
