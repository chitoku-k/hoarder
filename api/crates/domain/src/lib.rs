#![allow(clippy::too_many_arguments)]

pub mod entity;
pub mod parser;
pub mod processor;
pub mod repository;
pub mod service;
