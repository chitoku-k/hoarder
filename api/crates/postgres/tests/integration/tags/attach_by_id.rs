use std::collections::BTreeSet;

use domain::{
    entity::tags::{AliasSet, Tag, TagDepth, TagId},
    error::ErrorKind,
    repository::tags::TagsRepository,
};
use chrono::{TimeZone, Utc};
use futures::TryStreamExt;
use postgres::tags::PostgresTagsRepository;
use pretty_assertions::{assert_eq, assert_matches};
use sqlx::Row;
use test_context::test_context;
use uuid::{uuid, Uuid};

use super::DatabaseContext;

#[test_context(DatabaseContext)]
#[tokio::test]
async fn with_depth_succeeds(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.attach_by_id(
        TagId::from(uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60")),
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagDepth::new(2, 2),
    ).await.unwrap();

    assert_eq!(actual.name, "七森中☆ごらく部".to_string());
    assert_eq!(actual.kana, "ななもりちゅうごらくぶ".to_string());
    assert_eq!(actual.aliases, AliasSet::default());
    assert_eq!(actual.parent, Some(Box::new(Tag {
        id: TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        name: "東方Project".to_string(),
        kana: "とうほうProject".to_string(),
        aliases: AliasSet::new(BTreeSet::from(["東方".to_string()])),
        parent: None,
        children: Vec::new(),
        created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 8).unwrap(),
        updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 10).unwrap(),
    })));
    assert_eq!(actual.children, vec![
        Tag {
            id: TagId::from(uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5")),
            name: "赤座あかり".to_string(),
            kana: "あかざあかり".to_string(),
            aliases: AliasSet::default(),
            parent: None,
            children: Vec::new(),
            created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 9).unwrap(),
            updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap(),
        },
        Tag {
            id: TagId::from(uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6")),
            name: "歳納京子".to_string(),
            kana: "としのうきょうこ".to_string(),
            aliases: AliasSet::default(),
            parent: None,
            children: Vec::new(),
            created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 7).unwrap(),
            updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap(),
        },
        Tag {
            id: TagId::from(uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707")),
            name: "船見結衣".to_string(),
            kana: "ふなみゆい".to_string(),
            aliases: AliasSet::default(),
            parent: None,
            children: Vec::new(),
            created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 7).unwrap(),
            updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 7).unwrap(),
        },
        Tag {
            id: TagId::from(uuid!("991a287c-e77d-456f-94b4-293334674d0e")),
            name: "吉川ちなつ".to_string(),
            kana: "よしかわちなつ".to_string(),
            aliases: AliasSet::default(),
            parent: None,
            children: Vec::new(),
            created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 8).unwrap(),
            updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 6).unwrap(),
        },
    ]);
    assert_eq!(actual.created_at, Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap());
    assert_eq!(actual.updated_at, Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap());

    let actual: Vec<_> = sqlx::query(r#"SELECT "ancestor_id", "descendant_id", "distance" FROM "tag_paths" WHERE "descendant_id" IN ($1, $2, $3, $4, $5, $6, $7) ORDER BY "descendant_id", "distance" DESC"#)
        .bind(uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"))
        .bind(uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"))
        .bind(uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"))
        .bind(uuid!("991a287c-e77d-456f-94b4-293334674d0e"))
        .bind(uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"))
        .bind(uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"))
        .bind(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"))
        .fetch(&ctx.pool)
        .try_collect()
        .await
        .unwrap();

    assert_eq!(actual.len(), 23);

    assert_eq!(actual[0].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[0].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[0].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[1].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[1].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[1].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[2].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[2].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[2].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[3].get::<Uuid, &str>("ancestor_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[3].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[3].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[4].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[4].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[4].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[5].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[5].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[5].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[6].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[6].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[6].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[7].get::<Uuid, &str>("ancestor_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[7].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[7].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[8].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[8].get::<Uuid, &str>("descendant_id"), uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"));
    assert_eq!(actual[8].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[9].get::<Uuid, &str>("ancestor_id"), uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"));
    assert_eq!(actual[9].get::<Uuid, &str>("descendant_id"), uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"));
    assert_eq!(actual[9].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[10].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[10].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[10].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[11].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[11].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[11].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[12].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[12].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[12].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[13].get::<Uuid, &str>("ancestor_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[13].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[13].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[14].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[14].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[14].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[15].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[15].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[15].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[16].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[16].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[16].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[17].get::<Uuid, &str>("ancestor_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[17].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[17].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[18].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[18].get::<Uuid, &str>("descendant_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[18].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[19].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[19].get::<Uuid, &str>("descendant_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[19].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[20].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[20].get::<Uuid, &str>("descendant_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[20].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[21].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[21].get::<Uuid, &str>("descendant_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[21].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[22].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[22].get::<Uuid, &str>("descendant_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[22].get::<i32, &str>("distance"), 0);
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn without_depth_succeeds(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.attach_by_id(
        TagId::from(uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60")),
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagDepth::new(0, 0),
    ).await.unwrap();

    assert_eq!(actual.name, "七森中☆ごらく部".to_string());
    assert_eq!(actual.kana, "ななもりちゅうごらくぶ".to_string());
    assert_eq!(actual.aliases, AliasSet::default());
    assert_eq!(actual.parent, None);
    assert_eq!(actual.children, Vec::new());
    assert_eq!(actual.created_at, Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap());
    assert_eq!(actual.updated_at, Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap());

    let actual: Vec<_> = sqlx::query(r#"SELECT "ancestor_id", "descendant_id", "distance" FROM "tag_paths" WHERE "descendant_id" IN ($1, $2, $3, $4, $5, $6, $7) ORDER BY "descendant_id", "distance" DESC"#)
        .bind(uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"))
        .bind(uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"))
        .bind(uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"))
        .bind(uuid!("991a287c-e77d-456f-94b4-293334674d0e"))
        .bind(uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"))
        .bind(uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"))
        .bind(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"))
        .fetch(&ctx.pool)
        .try_collect()
        .await
        .unwrap();

    assert_eq!(actual.len(), 23);

    assert_eq!(actual[0].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[0].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[0].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[1].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[1].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[1].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[2].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[2].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[2].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[3].get::<Uuid, &str>("ancestor_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[3].get::<Uuid, &str>("descendant_id"), uuid!("12c4101e-722f-4172-9fe2-7862ebbc8fc5"));
    assert_eq!(actual[3].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[4].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[4].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[4].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[5].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[5].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[5].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[6].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[6].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[6].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[7].get::<Uuid, &str>("ancestor_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[7].get::<Uuid, &str>("descendant_id"), uuid!("69c4860d-24d0-41f6-a3ab-ac07dea5abd6"));
    assert_eq!(actual[7].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[8].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[8].get::<Uuid, &str>("descendant_id"), uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"));
    assert_eq!(actual[8].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[9].get::<Uuid, &str>("ancestor_id"), uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"));
    assert_eq!(actual[9].get::<Uuid, &str>("descendant_id"), uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"));
    assert_eq!(actual[9].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[10].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[10].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[10].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[11].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[11].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[11].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[12].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[12].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[12].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[13].get::<Uuid, &str>("ancestor_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[13].get::<Uuid, &str>("descendant_id"), uuid!("991a287c-e77d-456f-94b4-293334674d0e"));
    assert_eq!(actual[13].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[14].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[14].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[14].get::<i32, &str>("distance"), 3);

    assert_eq!(actual[15].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[15].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[15].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[16].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[16].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[16].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[17].get::<Uuid, &str>("ancestor_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[17].get::<Uuid, &str>("descendant_id"), uuid!("a2a6c29d-18d0-47b1-a324-88e93c267707"));
    assert_eq!(actual[17].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[18].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[18].get::<Uuid, &str>("descendant_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[18].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[19].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[19].get::<Uuid, &str>("descendant_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[19].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[20].get::<Uuid, &str>("ancestor_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[20].get::<Uuid, &str>("descendant_id"), uuid!("e8d32062-0185-43e8-a27d-6ca707d7dd60"));
    assert_eq!(actual[20].get::<i32, &str>("distance"), 0);

    assert_eq!(actual[21].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[21].get::<Uuid, &str>("descendant_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[21].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[22].get::<Uuid, &str>("ancestor_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[22].get::<Uuid, &str>("descendant_id"), uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb"));
    assert_eq!(actual[22].get::<i32, &str>("distance"), 0);
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn root_fails(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.attach_by_id(
        TagId::from(uuid!("00000000-0000-0000-0000-000000000000")),
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagDepth::new(0, 0),
    ).await.unwrap_err();

    assert_matches!(actual.kind(), ErrorKind::TagAttachingRoot);
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn root_parent_fails(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.attach_by_id(
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagId::from(uuid!("00000000-0000-0000-0000-000000000000")),
        TagDepth::new(0, 0),
    ).await.unwrap_err();

    assert_matches!(actual.kind(), ErrorKind::TagAttachingRoot);
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn self_fails(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.attach_by_id(
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagDepth::new(0, 0),
    ).await.unwrap_err();

    assert_matches!(actual.kind(), ErrorKind::TagAttachingToItself { id } if id == &TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")));
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn descendant_fails(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.attach_by_id(
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagId::from(uuid!("7648d9b5-e0f0-48c2-870c-1fcd60a099de")),
        TagDepth::new(0, 0),
    ).await.unwrap_err();

    assert_matches!(actual.kind(), ErrorKind::TagAttachingToDescendant { id } if id == &TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")));
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn non_existing_fails(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.attach_by_id(
        TagId::from(uuid!("11111111-1111-1111-1111-111111111111")),
        TagId::from(uuid!("fe81a56d-165b-446d-aebb-ca59e5acf3cb")),
        TagDepth::new(0, 0),
    ).await.unwrap_err();

    assert_matches!(actual.kind(), ErrorKind::TagNotFound { id } if id == &TagId::from(uuid!("11111111-1111-1111-1111-111111111111")));
}
