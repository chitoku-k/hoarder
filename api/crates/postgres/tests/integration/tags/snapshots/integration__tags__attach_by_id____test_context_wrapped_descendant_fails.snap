---
source: crates/postgres/tests/integration/tags/attach_by_id.rs
expression: ctx.queries()
---
[[queries]]
sql = '''
SELECT
  COUNT(*)
FROM
  "tag_paths"
WHERE
  "ancestor_id" = $1
  AND "descendant_id" = $2'''
rows_affected = 1
rows_returned = 1
